src/signing/petition_types/petition_entity.rs

```diff
@@ -0,0 +1,557 @@
+                    .collect::<IndexSet<T>>()
+            },
+        )
+    }
+
+    pub fn all_factor_instances(&self) -> IndexSet<OwnedFactorInstance> {
+        self.union_of(|l| l.factor_instances())
+            .into_iter()
+            .map(|f| OwnedFactorInstance::owned_factor_instance(self.entity.clone(), f.clone()))
+            .collect::<IndexSet<_>>()
+    }
+
+    pub fn all_skipped_factor_instance(&self) -> IndexSet<HierarchicalDeterministicFactorInstance> {
+        self.union_of(|f| f.all_skipped())
+    }
+
+    pub fn all_skipped_factor_sources(&self) -> IndexSet<FactorSourceIDFromHash> {
+        self.all_skipped_factor_instance()
+            .into_iter()
+            .map(|f| f.factor_source_id)
+            .collect::<IndexSet<_>>()
+    }
+
+    pub fn all_signatures(&self) -> IndexSet<HDSignature> {
+        self.union_of(|f| f.all_signatures())
+    }
+
+    fn with_list<F, T>(list: &Option<RefCell<PetitionFactors>>, map: F) -> Option<T>
+    where
+        F: Fn(&PetitionFactors) -> T,
+    {
+        list.as_ref().map(|refcell| map(&refcell.borrow()))
+    }
+
+    fn on_list<F, R>(&self, kind: FactorListKind, r#do: &F) -> Option<R>
+    where
+        F: Fn(&PetitionFactors) -> R,
+    {
+        match kind {
+            FactorListKind::Threshold => Self::with_list(&self.threshold_factors, r#do),
+            FactorListKind::Override => Self::with_list(&self.override_factors, r#do),
+        }
+    }
+
+    fn both<F, C, T, R>(&self, r#do: F, combine: C) -> R
+    where
+        F: Fn(&PetitionFactors) -> T,
+        C: Fn(Option<T>, Option<T>) -> R,
+    {
+        let t = self.on_list(FactorListKind::Threshold, &r#do);
+        let o = self.on_list(FactorListKind::Override, &r#do);
+        combine(t, o)
+    }
+
+    fn both_void<F, R>(&self, r#do: F)
+    where
+        F: Fn(&PetitionFactors) -> R,
+    {
+        self.both(r#do, |_, _| ())
+    }
+
+    pub fn skipped_factor_source_if_relevant(&self, factor_source_id: &FactorSourceIDFromHash) {
+        self.both_void(|l| l.skip_if_references(factor_source_id, true));
+    }
+
+    /// # Panics
+    /// Panics if this factor source has already been skipped or signed with.
+    ///
+    /// Or panics if the factor source is not known to this petition.
+    pub fn add_signature(&self, signature: HDSignature) {
+        self.both(|l| l.add_signature_if_relevant(&signature), |t, o| {
+            match (t, o) {
+                (Some(true), Some(true)) => {
+                    unreachable!("Matrix of FactorInstances does not allow for a factor to be present in both threshold and override list, thus this will never happen.")
+                }
+                (Some(false), Some(false)) => panic!("Factor source not found in any of the lists."),
+                (None, None) => panic!("Programmer error! Must have at least one factors list."), 
+                _ => (),
+            }
+        })
+    }
+
+    pub fn invalid_transactions_if_skipped_factors(
```

Comment:
Why `transactions`? the petition has only one intent hash

Reply:
Why indeed! I switched to `Option` instead of working on empty sets:

Fixed in: https://github.com/radixdlt/sargon-mfa/pull/6/commits/b72ca86a17df3bf1c5e53dbe232a084b62241f3e